core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/allowances/mareco_impl/AcceleratingSlopeCoast.java

```diff
@@ -91,61 +141,63 @@ public static ArrayList<AcceleratingSlopeCoast> findAll(
                 continue;
             }
 
+            double previousPosition = NaN;
+            double previousAcceleration = NaN;
+            AcceleratingSlopeCoast currentAcceleratingSlope = null;
+
             // constant variables for this plateau
-            double rollingResistance = context.rollingStock.getRollingResistance(speed);
             var envelopePart = cursor.getPart();
             var positionStep = context.timeStep * speed;
 
-            // initializing variables
-            double position = cursor.getPosition();
-            double weightForce = TrainPhysicsIntegrator.getWeightForce(context.rollingStock, context.path, position);
-            var naturalAcceleration = TrainPhysicsIntegrator.computeAcceleration(context.rollingStock,
-                    rollingResistance, weightForce, speed, 0, 0, 1);
-
-            while (cursor.getPosition() <= envelopePart.getEndPos()) {
+            while (!cursor.hasReachedEnd() && cursor.getPosition() <= envelopePart.getEndPos()) {
+                double position = cursor.getPosition();
+                double naturalAcceleration = getNaturalAcceleration(context, position, speed);
                 if (naturalAcceleration > 0) {
-                    // beginning of accelerating slope
-                    // add acceleration * distance step to the weighted mean acceleration
-                    var distanceStep = cursor.getStepLength();
-                    meanAcceleration += naturalAcceleration * distanceStep;
-                    accelerationLength += distanceStep;
-                    if (previousAcceleration < 0)
-                        currentAcceleratingSlope.previousAccelerationEstimate = previousAcceleration;
-                } else if (naturalAcceleration <= 0 && accelerationLength > 0) {
+                    // Accelerating slope
+                    if (currentAcceleratingSlope == null)
+                        currentAcceleratingSlope = new AcceleratingSlopeCoast(position, speed);
+                    currentAcceleratingSlope.meanAccelerationBuilder.addSegment(positionStep, naturalAcceleration);
+                } else if (currentAcceleratingSlope != null) {
                     // end the accelerating slope
-                    currentAcceleratingSlope.endPos = previousPosition;
-                    currentAcceleratingSlope.speedLimit = speed;
-                    meanAcceleration /= accelerationLength;
-                    currentAcceleratingSlope.slopeAverageAcceleration = meanAcceleration;
-                    res.add(currentAcceleratingSlope);
-                    // reset meanAcceleration, accelerationLength, and accelerating slope
-                    meanAcceleration = 0;
-                    accelerationLength = 0;
-                    currentAcceleratingSlope = new AcceleratingSlopeCoast(0, 0, 0, 0);
+                    var endPos = interpolateAccelerationSignChange(
```

Comment:
I don't know if there's a way improve it but you're looking for the `acceleration = 0` spot twice : here and in `findExactStartAcceleratingSlope`. Do you think it would be possible to refactor this to call this method only once ?

Reply:
I don't think it would really work, we don't compute it in the same way, there's a lot more stuff going on in this loop.

Reply:
Ok